[distributed] Skip ProcessGroupNCCLTest if CUDA is not available

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28393 [distributed] Skip ProcessGroupNCCLTest if CUDA is not available**

We should skip this test if CUDA is not available and alert the user.
Previously, if this test was ran on cpu it would fail with:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  cuda runtime error (3) : This binary is linked with CUDA lazy stubs and underlying .so files were not loaded. CUDA functionality is disabled. Set env variable CUDA_LAZY_DEBUG to get messages during startup
```

Open question: should this file eventually be refactored to use `gtest` like we do for most other C++ tests?


Differential Revision: [D18054369](https://our.internmc.facebook.com/intern/diff/D18054369/)

rohan-varma committed:
[distributed] Skip ProcessGroupNCCLTest if CUDA is not available

We should skip this test if CUDA is not available and alert the user.
Previously, if this test was ran on cpu it would fail with:

Differential Revision: [D18054369](https://our.internmc.facebook.com/intern/diff/D18054369/)

[ghstack-poisoned]

mrshenli reviewed (approved):
Stamped. Besides adding this, I would assume there is some config to tell the internal test infra to only run this on GPU server? 

cc @yf225

rohan-varma commented:
@pietern Would be great if you could take a look at this too

pritamdamania87 commented on torch/lib/c10d/test/ProcessGroupNCCLTest.cpp:
```diff
@@ -414,6 +414,11 @@ void testReduceScatter(const std::string& path, int rank, int size) {
 }
 
 int main(int argc, char** argv) {
+  if (!(c10::cuda::device_count() > 0)) {
```
nit: `if (!c10::cuda::is_available())`

rohan-varma commented on torch/lib/c10d/test/ProcessGroupNCCLTest.cpp:
```diff
@@ -414,6 +414,11 @@ void testReduceScatter(const std::string& path, int rank, int size) {
 }
 
 int main(int argc, char** argv) {
+  if (!(c10::cuda::device_count() > 0)) {
```
@pritamdamania87 this file doesn't use gtest right now, it should probably be rewritten to use gtest if we want to do this. For now, I've just changed this to `if (!c10::cuda::is_available())`. I've created a gh issue to track this: https://github.com/pytorch/pytorch/issues/28410

pietern replied:
`GTEST_SKIP` is not available in the googletest version we use (1.8 AFAIK). Last time I checked it is only available on master (unreleased). We use googletest from a submodule so we could bump it to bring it in.

rohan-varma replied:
I bumped the googletest to 1.10.0 which has `GTEST_SKIP`. The PR is here: https://github.com/pytorch/pytorch/pull/28432

rohan-varma committed:
Update on "[distributed] Skip ProcessGroupNCCLTest if CUDA is not available"


We should skip this test if CUDA is not available and alert the user.
Previously, if this test was ran on cpu it would fail with:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  cuda runtime error (3) : This binary is linked with CUDA lazy stubs and underlying .so files were not loaded. CUDA functionality is disabled. Set env variable CUDA_LAZY_DEBUG to get messages during startup
```

Open question: should this file eventually be refactored to use `gtest` like we do for most other C++ tests?

Differential Revision: [D18054369](https://our.internmc.facebook.com/intern/diff/D18054369/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] Skip ProcessGroupNCCLTest if CUDA is not available"


We should skip this test if CUDA is not available and alert the user.
Previously, if this test was ran on cpu it would fail with:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  cuda runtime error (3) : This binary is linked with CUDA lazy stubs and underlying .so files were not loaded. CUDA functionality is disabled. Set env variable CUDA_LAZY_DEBUG to get messages during startup
```

Open question: should this file eventually be refactored to use `gtest` like we do for most other C++ tests?

Differential Revision: [D18054369](https://our.internmc.facebook.com/intern/diff/D18054369/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] Skip ProcessGroupNCCLTest if CUDA is not available"


We should skip this test if CUDA is not available and alert the user.
Previously, if this test was ran on cpu it would fail with:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  cuda runtime error (3) : This binary is linked with CUDA lazy stubs and underlying .so files were not loaded. CUDA functionality is disabled. Set env variable CUDA_LAZY_DEBUG to get messages during startup
```

Open question: should this file eventually be refactored to use `gtest` like we do for most other C++ tests?

Differential Revision: [D18054369](https://our.internmc.facebook.com/intern/diff/D18054369/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] Skip ProcessGroupNCCLTest if CUDA is not available"


We should skip this test if CUDA is not available and alert the user.
Previously, if this test was ran on cpu it would fail with:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  cuda runtime error (3) : This binary is linked with CUDA lazy stubs and underlying .so files were not loaded. CUDA functionality is disabled. Set env variable CUDA_LAZY_DEBUG to get messages during startup
```

Open question: should this file eventually be refactored to use `gtest` like we do for most other C++ tests?

Differential Revision: [D18054369](https://our.internmc.facebook.com/intern/diff/D18054369/)

[ghstack-poisoned]

pietern commented on torch/lib/c10d/test/ProcessGroupNCCLTest.cpp:
```diff
@@ -414,6 +414,11 @@ void testReduceScatter(const std::string& path, int rank, int size) {
 
 int main(int argc, char** argv) {
+  if (!at::cuda::is_available()) {
+    // skip the test
```
Comment doesn't add anything and can go IMO.

pritamdamania87 commented on torch/lib/c10d/test/ProcessGroupNCCLTest.cpp:
```diff
@@ -414,6 +414,11 @@ void testReduceScatter(const std::string& path, int rank, int size) {
 int main(int argc, char** argv) {
+  if (!at::cuda::is_available()) {
+    // skip the test
+    std::cout << "CUDA not available, skipping test" << std::endl;
```
Is it possible to use glog here?

rohan-varma replied:
It doesn't seem like this is used, but looks like we can do `LOG(INFO) << message` which seems to be the pattern in the codebase.

rohan-varma replied:
Changed it to use `LOG(INFO) << ....`

rohan-varma committed:
Update on "[distributed] Skip ProcessGroupNCCLTest if CUDA is not available"


We should skip this test if CUDA is not available and alert the user.
Previously, if this test was ran on cpu it would fail with:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  cuda runtime error (3) : This binary is linked with CUDA lazy stubs and underlying .so files were not loaded. CUDA functionality is disabled. Set env variable CUDA_LAZY_DEBUG to get messages during startup
```

Open question: should this file eventually be refactored to use `gtest` like we do for most other C++ tests?

(Also includes a refactor of `ProcessGroupNCCLErrorsTest` to use `GTEST_SKIP`)

Differential Revision: [D18054369](https://our.internmc.facebook.com/intern/diff/D18054369/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] Skip ProcessGroupNCCLTest if CUDA is not available"


We should skip this test if CUDA is not available and alert the user.
Previously, if this test was ran on cpu it would fail with:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  cuda runtime error (3) : This binary is linked with CUDA lazy stubs and underlying .so files were not loaded. CUDA functionality is disabled. Set env variable CUDA_LAZY_DEBUG to get messages during startup
```

Open question: should this file eventually be refactored to use `gtest` like we do for most other C++ tests?


Differential Revision: [D18054369](https://our.internmc.facebook.com/intern/diff/D18054369/)

[ghstack-poisoned]
